pkg/plugin/plugin.go

```diff
@@ -0,0 +1,354 @@
+			response.Error = err
+			return response
+		}
+		page, err := getNextPageRes.Parse()
+		if err != nil {
+			log.DefaultLogger.Error("Error parsing Astarte page data", "error", err)
+			response.Error = err
+			return response
+		}
+
+		// Again assuming only individually aggregated interfaces are supported
+		data, ok := page.([]client.DatastreamIndividualValue)
+		if !ok {
+			response.Error = fmt.Errorf("Could not find individual datastream values for device %s on interface %s, path %s", qm.Device, qm.InterfaceName, qm.Path)
+			log.DefaultLogger.Error("Error on value type read", "error", response.Error)
+			return response
+		}
+
+		log.DefaultLogger.Debug("Start reading Astarte data")
+
+		for _, v := range data {
+			switch v.Value.(type) {
+			case float64:
+				timestamps = append(timestamps, v.Timestamp)
+				values = append(values, v.Value.(float64))
+			case int64:
+				timestamps = append(timestamps, v.Timestamp)
+				values = append(values, float64(v.Value.(int64)))
+			case string:
+				if f, err := strconv.ParseFloat(v.Value.(string), 64); err != nil {
+					log.DefaultLogger.Warn("Could not parse as numeric datatype", "value", v.Value, "error", err)
+				} else {
+					timestamps = append(timestamps, v.Timestamp)
+					values = append(values, f)
+				}
+			default:
+				response.Error = fmt.Errorf("Device %s has data of non-numeric type on interface %s, path %s", qm.Device, qm.InterfaceName, qm.Path)
+				log.DefaultLogger.Error("Error on value type read", "error", response.Error)
+				return response
+			}
+		}
+	}
+
+	log.DefaultLogger.Debug("Successful Astarte data read")
+
+	TimeField := data.NewField("Time", nil, timestamps)
+	ValueField := data.NewField("Value", nil, values)
+	frame.Fields = append(frame.Fields, TimeField, ValueField)
+	response.Frames = append(response.Frames, frame)
+	return response
+}
+
+// CheckHealth handles health checks sent from Grafana to the plugin.
+// The main use case for these health checks is the test button on the
+// datasource configuration page which allows users to verify that
+// a datasource is working as expected.
+func (d *AppEngineDatasource) CheckHealth(_ context.Context, req *backend.CheckHealthRequest) (*backend.CheckHealthResult, error) {
+	log.DefaultLogger.Debug("CheckHealth called", "request", req)
+
+	var status = backend.HealthStatusOk
+	var message = "Data source is working"
+
+	// Run an actual query to Astarte, so that our JWT is checked, too
+	deviceStatsCall, _ := d.astarteAPIClient.GetDevicesStats(d.realm)
+
+	log.DefaultLogger.Debug("stats request", "curl", deviceStatsCall.ToCurl(d.astarteAPIClient))
+
+	_, err := deviceStatsCall.Run(d.astarteAPIClient)
+	if err != nil {
+		log.DefaultLogger.Error("CheckHealth error", "err", err)
+		status = backend.HealthStatusError
+		message = err.Error()
+	}
+
+	log.DefaultLogger.Debug("CheckHealth response")
+
+	return &backend.CheckHealthResult{
+		Status:  status,
+		Message: message,
+	}, nil
```

Comment:
I don't like the variable definition at the beginning of the method.
What about
```suggestion
	// Run an actual query to Astarte, so that our JWT is checked, too
	deviceStatsCall, _ := d.astarteAPIClient.GetDevicesStats(d.realm)

	log.DefaultLogger.Debug("stats request", "curl", deviceStatsCall.ToCurl(d.astarteAPIClient))

	_, err := deviceStatsCall.Run(d.astarteAPIClient)
	if err != nil {
		log.DefaultLogger.Error("CheckHealth error", "err", err)
		return &backend.CheckHealthResult{
		        Status:  backend.HealthStatusError
		        Message: err.Error(),
	        }, nil
	}

	log.DefaultLogger.Debug("CheckHealth response")

	return &backend.CheckHealthResult{
		Status:  backend.HealthStatusOk,
		Message: "Data source is working",
	}, nil
```
Also, the method signature tells us that if the healthCheck fails we should receive an error. However, we inform that an error occurred by setting the `CheckHealthResult.Message` field, while the error is `nil` regardless.

Reply:
It makes sense to return `err` for the health check, as errors in communication between Astarte and plugin backend and between plugin backend and frontend are both health errors. In other cases, you want to differentiate the two kinds of errors (e.g. an Astarte error should be displayed in the plugin frontend, too).